llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -265,6 +275,69 @@ void collectCompositeElementsInfoRecursive(
   }
 }
 
+/// Recursively iterates over a composite type in order to collect information
+/// about default values of its scalar elements.
+void collectCompositeElementsDefaultValuesRecursive(
+    const Module &M, Constant *C, unsigned &Offset,
```

Comment:
NIT: processing of composite spec constants is similar in this function, `emitSpecConstantRecursiveImpl` and `generateSpecConstDefaultValueMetadata`. Could be factored out using e.g. the visitor pattern. At least a TODO for now would be good.